File: backend/app/schemas/trazabilidad.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.models.produccion import DestinoCosecha, EstadoCampana
from app.models.trazabilidad import EstadoSanitarioAnalisis, OrigenAnalisis
from app.schemas.produccion import (
    CicloCampanaResponse,
    RegistroCosechaResponse,
    RegistroFitosanitarioResponse,
    RegistroRiegoResponse,
    RegistroTrabajoResponse,
)


# ── Foto ────────────────────────────────────────────────────────────────────

class FotoBase(BaseModel):
    parcela_id: str
    fecha: date
    categoria: str
    descripcion: str | None = None


class FotoCreate(FotoBase):
    pass


class FotoResponse(FotoBase):
    id: str
    url: str
    created_by: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Analisis de Calidad ───────────────────────────────────────────────────────

class AnalisisCalidadBase(BaseModel):
    parcela_id: str
    fecha: date
    origen: OrigenAnalisis
    brix: float | None = None
    acidez: float | None = None
    ph: float | None = None
    estado_sanitario: EstadoSanitarioAnalisis | None = None
    laboratorio_nombre: str | None = None
    observaciones: str | None = None


class AnalisisCalidadCreate(AnalisisCalidadBase):
    pass


class AnalisisCalidadResponse(AnalisisCalidadBase):
    id: str
    informe_url: str | None = None
    created_by: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Historial / Ficha de Trazabilidad ─────────────────────────────────────────

class ComplianceFitosanitario(BaseModel):
    fitosanitario_id: str
    fecha_aplicacion: date
    producto_nombre: str
    fecha_habilitacion_cosecha: date
    estado: Literal["cumplido", "incumplido", "pendiente"]
    cosecha_conflictiva_id: str | None = None
    cosecha_conflictiva_fecha: date | None = None


class CumplimientoEstadoItem(BaseModel):
    estado_campana: EstadoCampana
    estado_campana_label: str
    fecha_inicio: date
    fecha_fin: date
    riegos_esperados: int
    mm_aplicados: float
    riegos_equivalentes: float
    cumplimiento_pct: float
    cumplido: bool


class ResumenDestinoItem(BaseModel):
    destino: DestinoCosecha
    destino_label: str
    kg_total: float
    n_registros: int
    pct_del_total: float


class CentroideItem(BaseModel):
    lat: float
    lng: float


class TareaResumenItem(BaseModel):
    tarea: str
    unidad_medida_label: str
    fecha_inicio: date
    fecha_fin: date
    registros: int


# ── Enlace publico (Fase 3) ──────────────────────────────────────────────────

class EnlacePublicoCreate(BaseModel):
    desde: date
    hasta: date


class EnlacePublicoResponse(BaseModel):
    id: str
    parcela_id: str
    token: str
    desde: date
    hasta: date
    activo: bool
    created_at: datetime
    revoked_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


# Vista publica -- schemas curados por allow-list explicito. NO se reusan los
# schemas internos (RegistroFitosanitarioResponse etc.) porque heredan
# `responsable`/`comprador`/`trabajador_id` de sus *Base: un campo nuevo
# agregado a futuro a esos schemas se filtraria solo. Aca cada campo esta
# listado a mano.

class FitosanitarioPublicoItem(BaseModel):
    fecha: date
    producto_nombre: str
    dosis_lt_ha: float
    dias_carencia: int
    fecha_habilitacion_cosecha: date
    dias_reingreso: int
    fecha_habilitacion_reingreso: date
    estado_compliance: Literal["cumplido", "incumplido", "pendiente"]


class FotoPublicaItem(BaseModel):
    url: str
    categoria: str
    fecha: date
    descripcion: str | None = None


class AnalisisPublicoItem(BaseModel):
    fecha: date
    origen: OrigenAnalisis
    brix: float | None = None
    acidez: float | None = None
    ph: float | None = None
    estado_sanitario: EstadoSanitarioAnalisis | None = None
    laboratorio_nombre: str | None = None
    informe_url: str | None = None


class EmpresaItem(BaseModel):
    razon_social: str
    cuit: str
    domicilio: str


class ResumenPublicoItem(BaseModel):
    kg_total: float
    meta_produccion_kg: float | None = None
    pct_meta_produccion: float | None = None
    mm_riego_total: float
    mm_objetivo_anual: float
    pct_objetivo_riego: float | None = None
    horas_de_frio: float | None = None
    fitos_cumplidos: int
    fitos_pendientes: int
    fitos_incumplidos: int


class HistorialPublicoResponse(BaseModel):
    parcela_nombre: str
    parcela_tipo: str
    parcela_variedad: str | None = None
    parcela_variedad_descripcion: str | None = None
    parcela_superficie_ha: float | None = None
    parcela_finca: str | None = None
    parcela_tipo_riego: str | None = None
    parcela_cobertura_invierno: str | None = None
    parcela_centroide: CentroideItem | None = None
    desde: date
    hasta: date
    resumen: ResumenPublicoItem
    fitosanitarios: list[FitosanitarioPublicoItem]
    cumplimiento_riego_por_estado: list[CumplimientoEstadoItem]
    resumen_destino: list[ResumenDestinoItem]
    tareas_resumen: list[TareaResumenItem]
    fotos: list[FotoPublicaItem]
    analisis_calidad: list[AnalisisPublicoItem]
    empresa: EmpresaItem


class HistorialParcelaResponse(BaseModel):
    parcela_id: str
    parcela_nombre: str
    desde: date
    hasta: date
    riegos: list[RegistroRiegoResponse]
    fitosanitarios: list[RegistroFitosanitarioResponse]
    trabajos: list[RegistroTrabajoResponse]
    cosechas: list[RegistroCosechaResponse]
    ciclos_campana: list[CicloCampanaResponse]
    fotos: list[FotoResponse]
    analisis_calidad: list[AnalisisCalidadResponse]
    compliance_fitosanitarios: list[ComplianceFitosanitario]
    parcela_variedad_descripcion: str | None = None
    parcela_centroide: CentroideItem | None = None
    parcela_tipo_riego: str | None = None
    parcela_cobertura_invierno: str | None = None
    cumplimiento_riego_por_estado: list[CumplimientoEstadoItem]
    resumen_destino: list[ResumenDestinoItem]
    tareas_resumen: list[TareaResumenItem]
    horas_de_frio: float | None = None
    mm_objetivo_anual: float
    meta_produccion_kg: float | None = None

File: frontend/lib/api/trazabilidad.ts
import api from '@/lib/api'
import type { RiegoResponse } from '@/lib/api/riego'
import type { FitosanitarioResponse } from '@/lib/api/fitosanitarios'
import type { RegistroTrabajoResponse } from '@/lib/api/produccion'
import type { RegistroCosechaResponse } from '@/lib/api/cosecha'
import type { CicloCampanaItem } from '@/lib/api/produccion'

// ── Fotos ────────────────────────────────────────────────────────────────────

export interface FotoResponse {
  id: string
  parcela_id: string
  fecha: string
  categoria: string
  descripcion: string | null
  url: string
  created_by: string
  created_at: string
}

export async function createFoto(payload: {
  parcela_id: string
  fecha: string
  categoria: string
  descripcion?: string
  file: File
}): Promise<FotoResponse> {
  const formData = new FormData()
  formData.append('parcela_id', payload.parcela_id)
  formData.append('fecha', payload.fecha)
  formData.append('categoria', payload.categoria)
  if (payload.descripcion) formData.append('descripcion', payload.descripcion)
  formData.append('file', payload.file)
  const { data } = await api.post('/trazabilidad/fotos/', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function getFotos(params: {
  parcela_id?: string; desde?: string; hasta?: string
}): Promise<FotoResponse[]> {
  const { data } = await api.get('/trazabilidad/fotos/', { params })
  return data
}

export async function deleteFoto(id: string): Promise<void> {
  await api.delete(`/trazabilidad/fotos/${id}`)
}

// ── Analisis de Calidad ───────────────────────────────────────────────────────

export type OrigenAnalisis = 'propio' | 'laboratorio'
export type EstadoSanitarioAnalisis = 'sano' | 'con_observaciones' | 'rechazado'

export const ORIGEN_ANALISIS_LABELS: Record<OrigenAnalisis, string> = {
  propio: 'Medición propia',
  laboratorio: 'Informe de laboratorio',
}

export const ESTADO_SANITARIO_LABELS: Record<EstadoSanitarioAnalisis, string> = {
  sano: 'Sano',
  con_observaciones: 'Con observaciones',
  rechazado: 'Rechazado',
}

export interface AnalisisCalidadResponse {
  id: string
  parcela_id: string
  fecha: string
  origen: OrigenAnalisis
  brix: number | null
  acidez: number | null
  ph: number | null
  estado_sanitario: EstadoSanitarioAnalisis | null
  laboratorio_nombre: string | null
  informe_url: string | null
  observaciones: string | null
  created_by: string
  created_at: string
}

export async function createAnalisisCalidad(payload: {
  parcela_id: string
  fecha: string
  origen: OrigenAnalisis
  brix?: number
  acidez?: number
  ph?: number
  estado_sanitario?: EstadoSanitarioAnalisis
  laboratorio_nombre?: string
  observaciones?: string
  file?: File
}): Promise<AnalisisCalidadResponse> {
  const formData = new FormData()
  formData.append('parcela_id', payload.parcela_id)
  formData.append('fecha', payload.fecha)
  formData.append('origen', payload.origen)
  if (payload.brix !== undefined) formData.append('brix', String(payload.brix))
  if (payload.acidez !== undefined) formData.append('acidez', String(payload.acidez))
  if (payload.ph !== undefined) formData.append('ph', String(payload.ph))
  if (payload.estado_sanitario) formData.append('estado_sanitario', payload.estado_sanitario)
  if (payload.laboratorio_nombre) formData.append('laboratorio_nombre', payload.laboratorio_nombre)
  if (payload.observaciones) formData.append('observaciones', payload.observaciones)
  if (payload.file) formData.append('file', payload.file)
  const { data } = await api.post('/trazabilidad/analisis/', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function getAnalisisCalidad(params: {
  parcela_id?: string; desde?: string; hasta?: string
}): Promise<AnalisisCalidadResponse[]> {
  const { data } = await api.get('/trazabilidad/analisis/', { params })
  return data
}

export async function deleteAnalisisCalidad(id: string): Promise<void> {
  await api.delete(`/trazabilidad/analisis/${id}`)
}

// ── Historial / Ficha de Trazabilidad ─────────────────────────────────────────

export type ComplianceEstado = 'cumplido' | 'incumplido' | 'pendiente'

export const COMPLIANCE_ESTADO_LABELS: Record<ComplianceEstado, string> = {
  cumplido: 'Cumplido',
  incumplido: 'Incumplido',
  pendiente: 'Pendiente',
}

export interface ComplianceFitosanitarioItem {
  fitosanitario_id: string
  fecha_aplicacion: string
  producto_nombre: string
  fecha_habilitacion_cosecha: string
  estado: ComplianceEstado
  cosecha_conflictiva_id: string | null
  cosecha_conflictiva_fecha: string | null
}

export interface CumplimientoEstadoItem {
  estado_campana: string
  estado_campana_label: string
  fecha_inicio: string
  fecha_fin: string
  riegos_esperados: number
  mm_aplicados: number
  riegos_equivalentes: number
  cumplimiento_pct: number
  cumplido: boolean
}

export interface ResumenDestinoItem {
  destino: string
  destino_label: string
  kg_total: number
  n_registros: number
  pct_del_total: number
}

export interface CentroideItem {
  lat: number
  lng: number
}

export interface TareaResumenItem {
  tarea: string
  unidad_medida_label: string
  fecha_inicio: string
  fecha_fin: string
  registros: number
}

export interface HistorialParcelaResponse {
  parcela_id: string
  parcela_nombre: string
  desde: string
  hasta: string
  riegos: RiegoResponse[]
  fitosanitarios: FitosanitarioResponse[]
  trabajos: RegistroTrabajoResponse[]
  cosechas: RegistroCosechaResponse[]
  ciclos_campana: CicloCampanaItem[]
  fotos: FotoResponse[]
  analisis_calidad: AnalisisCalidadResponse[]
  compliance_fitosanitarios: ComplianceFitosanitarioItem[]
  parcela_variedad_descripcion: string | null
  parcela_centroide: CentroideItem | null
  parcela_tipo_riego: string | null
  parcela_cobertura_invierno: string | null
  cumplimiento_riego_por_estado: CumplimientoEstadoItem[]
  resumen_destino: ResumenDestinoItem[]
  tareas_resumen: TareaResumenItem[]
  horas_de_frio: number | null
  mm_objetivo_anual: number
  meta_produccion_kg: number | null
}

export async function getHistorialParcela(
  parcelaId: string, desde: string, hasta: string,
): Promise<HistorialParcelaResponse> {
  const { data } = await api.get(`/trazabilidad/parcela/${parcelaId}/historial`, {
    params: { desde, hasta },
  })
  return data
}

export async function downloadCartaPdf(
  parcelaId: string, desde: string, hasta: string,
): Promise<Blob> {
  const { data } = await api.get(`/trazabilidad/parcela/${parcelaId}/carta-pdf`, {
    params: { desde, hasta },
    responseType: 'blob',
  })
  return data
}

// ── Enlace público (Fase 3: QR sin login) ────────────────────────────────────

export interface EnlacePublicoResponse {
  id: string
  parcela_id: string
  token: string
  desde: string
  hasta: string
  activo: boolean
  created_at: string
  revoked_at: string | null
}

export async function crearEnlacePublico(
  parcelaId: string, desde: string, hasta: string,
): Promise<EnlacePublicoResponse> {
  const { data } = await api.post(`/trazabilidad/parcela/${parcelaId}/enlaces`, { desde, hasta })
  return data
}

export async function listEnlacesPublicos(parcelaId: string): Promise<EnlacePublicoResponse[]> {
  const { data } = await api.get(`/trazabilidad/parcela/${parcelaId}/enlaces`)
  return data
}

export async function revocarEnlacePublico(enlaceId: string): Promise<EnlacePublicoResponse> {
  const { data } = await api.post(`/trazabilidad/enlaces/${enlaceId}/revocar`)
  return data
}

// ── Vista pública (schema curado por el backend, sin responsable/comprador) ───

export interface FitosanitarioPublicoItem {
  fecha: string
  producto_nombre: string
  dosis_lt_ha: number
  dias_carencia: number
  fecha_habilitacion_cosecha: string
  dias_reingreso: number
  fecha_habilitacion_reingreso: string
  estado_compliance: ComplianceEstado
}

export interface FotoPublicaItem {
  url: string
  categoria: string
  fecha: string
  descripcion: string | null
}

export interface AnalisisPublicoItem {
  fecha: string
  origen: OrigenAnalisis
  brix: number | null
  acidez: number | null
  ph: number | null
  estado_sanitario: EstadoSanitarioAnalisis | null
  laboratorio_nombre: string | null
  informe_url: string | null
}

export interface ResumenPublicoItem {
  kg_total: number
  meta_produccion_kg: number | null
  pct_meta_produccion: number | null
  mm_riego_total: number
  mm_objetivo_anual: number
  pct_objetivo_riego: number | null
  horas_de_frio: number | null
  fitos_cumplidos: number
  fitos_pendientes: number
  fitos_incumplidos: number
}

export interface EmpresaItem {
  razon_social: string
  cuit: string
  domicilio: string
}

export interface HistorialPublicoResponse {
  parcela_nombre: string
  parcela_tipo: string
  parcela_variedad: string | null
  parcela_variedad_descripcion: string | null
  parcela_superficie_ha: number | null
  parcela_finca: string | null
  parcela_tipo_riego: string | null
  parcela_cobertura_invierno: string | null
  parcela_centroide: CentroideItem | null
  desde: string
  hasta: string
  resumen: ResumenPublicoItem
  fitosanitarios: FitosanitarioPublicoItem[]
  cumplimiento_riego_por_estado: CumplimientoEstadoItem[]
  resumen_destino: ResumenDestinoItem[]
  tareas_resumen: TareaResumenItem[]
  fotos: FotoPublicaItem[]
  analisis_calidad: AnalisisPublicoItem[]
  empresa: EmpresaItem
}

export async function getHistorialPublico(token: string): Promise<HistorialPublicoResponse> {
  const { data } = await api.get(`/trazabilidad/publica/${token}`)
  return data
}

export async function downloadCartaPdfPublica(token: string): Promise<Blob> {
  const { data } = await api.get(`/trazabilidad/publica/${token}/pdf`, { responseType: 'blob' })
  return data
}
